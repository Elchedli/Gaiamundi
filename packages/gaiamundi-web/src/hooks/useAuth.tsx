import React from 'react';
import {
  QueryObserverResult,
  RefetchOptions,
  UseMutateFunction,
  useQuery,
  useQueryClient,
} from 'react-query';

import { Alert } from 'components/Alert/Alert';
import { LoadingMessage } from 'components/Loader/LoadingMessage';
import { User } from 'interfaces/user';
import { strapi } from '../services/strapi';
import { useLocalStorageState } from './useLocalStorageState';

// const updateAccount = async (id: number, userData: UserSignUpFields) => {
//   return (await strapi.update(ContentType.USERS, id, userData)) as User;
// };

export interface AuthContextValue {
  user: User | undefined;
  isAuthenticated: boolean;
  setUser: (data: User | undefined) => void;
  authenticate: (user: User, jwt: string) => void;
  logout: UseMutateFunction;
  refetchUser: (
    options?: RefetchOptions | undefined
  ) => Promise<QueryObserverResult<User, Error>>;
  error: Error | null;
}

const AuthContext = React.createContext<AuthContextValue | null>(null);
AuthContext.displayName = 'AuthContext';

export interface AuthProviderProps {
  children: React.ReactNode;
}

export const AuthProvider = ({ children }: AuthProviderProps): JSX.Element => {
  const { value: jwtToken, persistValue: persistJwtToken } =
    useLocalStorageState('token');
  const queryClient = useQueryClient();
  const key = 'auth-user';

  const {
    data: user,
    error,
    isLoading,
    refetch,
  } = useQuery<User, Error>({
    queryKey: key,
    queryFn: async () => {
      return await strapi.currentUser(jwtToken);
    },
    enabled: !!jwtToken,
  });

  const setUser = (data: User | undefined) =>
    queryClient.setQueryData(key, data);

  const authenticate = (user: User, jwt: string) => {
    setUser(user);
    persistJwtToken(jwt);
  };

  const logout = () => {
    persistJwtToken('');
    queryClient.clear();
  };

  if (isLoading) {
    return <LoadingMessage />;
  }

  return (
    <AuthContext.Provider
      value={{
        user,
        isAuthenticated: !!jwtToken,
        error,
        setUser,
        authenticate,
        refetchUser: refetch,
        logout,
      }}
    >
      {error && <Alert>{error.message}</Alert>}
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => {
  const context = React.useContext(AuthContext);
  if (!context) {
    throw new Error(`useAuth must be used within an AuthProvider`);
  }
  return context;
};
