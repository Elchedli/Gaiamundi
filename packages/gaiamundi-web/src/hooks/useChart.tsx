import { CHART_TYPES } from 'interfaces/chart';
import { useMemo } from 'react';
import { useQuery } from 'react-query';
import { getChartById } from 'services/chart';

export const useChart = (chartId: number) => {
  const query = useQuery({
    queryKey: ['chart', chartId],
    queryFn: async () => await getChartById(chartId),
    enabled: !!chartId,
  });
  const chart = query.data?.data;
  const ChartComponent = useMemo(() => {
    return chart?.type
      ? CHART_TYPES[chart.type].ChartComponent
      : CHART_TYPES.column.ChartComponent;
  }, [chart]);

  return {
    ...query,
    ChartComponent,
  };
};
