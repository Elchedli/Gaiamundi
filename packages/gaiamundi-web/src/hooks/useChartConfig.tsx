import { LoadingMessage } from 'components/Loader/LoadingMessage';
import { Dimensions, Direction } from 'eazychart-core';
import { ColumnChart } from 'eazychart-react';
import { ApiData } from 'interfaces/api';
import {
  AllChartConfigProps as AllChartProps,
  CHART_TYPES,
  Chart,
  RawDatum,
  RawDatumType,
} from 'interfaces/chart';
import React, {
  FC,
  ReactNode,
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { useQuery, useQueryClient } from 'react-query';
import { getChartById } from 'services/chart';
import { getDataTypeMap, guessDomainKey } from 'utils/chart';
import { DEFAULT_DIMENSIONS, INITIAL_CHART_CONFIG } from 'utils/constants';

type ChartConfigContextType = {
  chart: ApiData<Chart>;
  ChartComponent: React.FC<AllChartProps>;
  updateChartProps: (_props: Partial<AllChartProps>) => void;
  updateChart: (config: Partial<Chart>) => void;
  rawData: RawDatum[];
  dataKeys: Record<string, RawDatumType>;
  dimensions: Dimensions;
  setDimensions: (_dimensions: Dimensions) => void;
};

const initialContext: ChartConfigContextType = {
  chart: INITIAL_CHART_CONFIG,
  ChartComponent: ColumnChart,
  updateChartProps: (_props: Partial<AllChartProps>) => undefined,
  updateChart: (_config: Partial<Chart>) => undefined,
  rawData: [],
  dataKeys: {},
  dimensions: DEFAULT_DIMENSIONS,
  setDimensions: (_dimensions: Dimensions) => undefined,
};

const ChartConfigContext =
  createContext<ChartConfigContextType>(initialContext);

type ChartConfigProviderProps = {
  chartId: number;
  rawData: RawDatum[];
  children: ReactNode;
};

export const ChartConfigProvider: FC<ChartConfigProviderProps> = ({
  chartId,
  rawData,
  children,
}) => {
  const [dimensions, setDimensions] = useState<Dimensions>(DEFAULT_DIMENSIONS);
  const queryClient = useQueryClient();
  const {
    data: response,
    error,
    isLoading,
  } = useQuery({
    queryKey: ['chart', chartId],
    queryFn: async () => await getChartById(chartId),
    enabled: !!chartId,
  });

  const chart = response?.data || INITIAL_CHART_CONFIG;
  const setChart = (newChart: ApiData<Chart> | undefined) => {
    queryClient.setQueryData(['chart', chartId], { data: newChart });
  };

  const updateChartProps = (props: Partial<AllChartProps>) => {
    setChart({
      ...chart,
      props: {
        ...chart.props,
        ...props,
      },
    });
  };

  const updateChart = (updates: Partial<Chart>) => {
    setChart({ ...chart, ...updates });
  };

  const ChartComponent = useMemo(() => {
    return chart?.type
      ? CHART_TYPES[chart.type].ChartComponent
      : CHART_TYPES.column.ChartComponent;
  }, [chart]);

  const dataKeys = useMemo(() => {
    if (rawData.length > 0) {
      return getDataTypeMap(rawData[0]);
    }
    return {};
  }, [rawData]);

  useEffect(() => {
    switch (chart.type) {
      case 'column':
        updateChartProps({
          xAxis: {
            ...chart.props.xAxis,
            domainKey: guessDomainKey(rawData, 'string', 0),
          },
          yAxis: {
            ...chart.props.yAxis,
            domainKey: guessDomainKey(rawData, 'number', 1),
          },
        });
        break;
      case 'lineColumn':
        updateChartProps({
          xAxis: {
            ...chart.props.xAxis,
            domainKey: guessDomainKey(rawData, 'string', 0),
          },
          yAxis: {
            ...chart.props.yAxis,
            domainKey: guessDomainKey(rawData, 'number', 1),
          },
        });
        break;
      case 'bar':
        updateChartProps({
          xAxis: {
            ...chart.props.xAxis,
            domainKey: guessDomainKey(rawData, 'number', 0),
          },
          yAxis: {
            ...chart.props.yAxis,
            domainKey: guessDomainKey(rawData, 'string', 1),
          },
        });
        break;
      case 'pie':
        updateChartProps({
          domainKey: guessDomainKey(rawData, 'number', 0),
        });
        break;
      case 'line':
        updateChartProps({
          xAxis: {
            ...chart.props.xAxis,
            domainKey: guessDomainKey(rawData, 'number', 0),
          },
          yAxis: {
            ...chart.props.yAxis,
            domainKey: guessDomainKey(rawData, 'number', 1),
          },
        });
        break;
      case 'area':
        updateChartProps({
          xAxis: {
            ...chart.props.xAxis,
            domainKey: guessDomainKey(rawData, 'number', 0),
          },
          yAxis: {
            ...chart.props.yAxis,
            domainKey: guessDomainKey(rawData, 'number', 1),
          },
        });
        break;
      case 'scatter':
        updateChartProps({
          xAxis: {
            ...chart.props.xAxis,
            domainKey: guessDomainKey(rawData, 'number', 0),
          },
          yAxis: {
            ...chart.props.yAxis,
            domainKey: guessDomainKey(rawData, 'number', 1),
          },
        });
        break;
      case 'bubble':
        updateChartProps({
          xAxis: {
            ...chart.props.xAxis,
            domainKey: guessDomainKey(rawData, 'number', 0),
          },
          yAxis: {
            ...chart.props.yAxis,
            domainKey: guessDomainKey(rawData, 'number', 1),
          },
        });
        break;

      default:
        break;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [chart.type, dataKeys]);

  if (isLoading) {
    return <LoadingMessage />;
  }

  if (error) {
    return <>{error}</>;
  }

  return (
    <ChartConfigContext.Provider
      value={{
        chart,
        updateChartProps,
        updateChart,
        ChartComponent,
        dataKeys,
        rawData,
        dimensions,
        setDimensions,
      }}
    >
      {children}
    </ChartConfigContext.Provider>
  );
};

export const useChart = () => {
  return useContext(ChartConfigContext);
};

export const useChartGridConfig = () => {
  const { chart, updateChartProps } = useChart();

  const updateGridDirection = (direction: Direction, isEnabled: boolean) => {
    let gridDirections = [...chart.props.grid.directions] as Direction[];
    if (isEnabled) {
      gridDirections.push(direction);
    } else {
      gridDirections = gridDirections.filter((dir) => dir !== direction);
    }
    updateChartProps({
      grid: {
        ...chart.props.grid,
        directions: gridDirections,
      },
    });
  };

  const updateGridColor = (color: string) => {
    updateChartProps({
      grid: {
        ...chart.props.grid,
        color,
      },
    });
  };

  return {
    grid: chart.props.grid,
    updateGridDirection,
    updateGridColor,
  };
};
