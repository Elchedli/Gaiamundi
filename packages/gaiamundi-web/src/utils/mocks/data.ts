import { ApiCollection, ApiData } from 'interfaces/api';
import { DataFragment } from 'interfaces/data-fragment';
import { Dataset } from 'interfaces/dataset';
import { UploadedFile } from 'interfaces/file';
import { GeoMap } from 'interfaces/geo-map';
import { Indicator } from 'interfaces/indicator';
import { PageCarto } from 'interfaces/page-carto';
import { Tag } from 'interfaces/tag';
import { User } from 'interfaces/user';
import mockMapJson from './map.json';

export const mockMapPath = mockMapJson;

export const mockFileUpload = {
  id: 184,
  name: 'exampleParsing2.csv',
  alternativeText: null,
  caption: null,
  width: null,
  height: null,
  formats: null,
  hash: 'example_Parsing2_ce8756a044',
  ext: '.csv',
  mime: 'text/csv',
  size: 0.05,
  url: '/uploads/example_Parsing2_ce8756a044.csv',
  previewUrl: null,
  provider: 'local',
  provider_metadata: null,
  createdAt: '2023-03-16T18:20:09.476Z',
  updatedAt: '2023-03-16T18:20:09.476Z',
};

export const mockUser: User = {
  id: 1,
  username: 'testuser',
  password: 'test',
  email: 'test@test.com',
  provider: 'local',
  confirmed: true,
  blocked: false,
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockGeoJson: ApiData<UploadedFile> = {
  id: 1,
  name: 'test.geojson',
  hash: 'abc123',
  provider: 'provider123',
  ext: '.geojson',
  mime: 'application/json',
  size: 1234,
  url: '/api/v1/geo-maps/id/thumbnail',
  formats: {
    thumbnail: {
      extension: 'ext',
      url: 'happyUrl',
      size: 360,
      hash: 'hashbar',
      mime: 'mimeExample',
    },
  },
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockCover: ApiData<UploadedFile> = {
  id: 3,
  name: 'image.png',
  hash: 'hashImage',
  provider: 'provider123',
  ext: '.png',
  mime: 'application/image',
  size: 1234,
  url: '/api/v1/geo-maps/id/thumbnailImage',
  formats: {
    small: {
      extension: 'ext',
      url: '/api/v1/geo-maps/id/thumbnailImage?720',
      size: 360,
      hash: 'hash720',
      mime: 'mimeExample',
    },
    thumbnail: {
      extension: 'ext',
      url: '/api/v1/geo-maps/id/thumbnailImage?720',
      size: 360,
      hash: 'hash720',
      mime: 'mimeExample',
    },
  },
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockGeoMap: ApiData<GeoMap> = {
  id: 1,
  name: 'Test Map',
  yearValidity: 2022,
  mesh: 'Test Maille',
  source: 'Test Source',
  license: 'Test License',
  properties: [
    {
      name: 'code',
      sample: '39',
      isGeoCode: true,
    },
    {
      name: 'nom',
      sample: 'Jura',
      isGeoCode: false,
    },
  ],
  owner: mockUser,
  geoJSON: mockGeoJson,
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockTags: ApiData<Tag>[] = [
  {
    id: 1,
    name: 'Tag A',
    count: 5,
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
  {
    id: 2,
    name: 'Tag B',
    count: 3,
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
  {
    id: 3,
    name: 'Tag C',
    count: 2,
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
];

export const mockDataset: ApiData<Dataset> = {
  id: 122,
  name: 'firstDataset',
  origin: 3,
  isPublic: true,
  owner: mockUser,
  csv: {
    id: 5,
    name: 'file.csv',
    hash: 'hashCSV',
    provider: 'provider123',
    ext: '.csv',
    mime: 'application/csv',
    size: 1234,
    url: '/api/v1/Uploads/id/csvFile.csv',
    formats: {
      thumbnail: {
        extension: 'ext',
        url: '/api/v1/Uploads/id/csvFile.csv?720',
        size: 360,
        hash: 'hash720',
        mime: 'mimeExample',
      },
    },
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockDataFragments: ApiData<DataFragment>[] = [
  {
    id: 1,
    name: 'FirstFragment',
    columns: [
      {
        name: 'firstColumn',
        sample: 5,
        source: 'source1',
        validity: '2021',
        isGeoCode: true,
      },
      {
        name: 'secondColumn',
        sample: 10.3,
        source: 'source2',
        validity: '2022',
        isGeoCode: false,
      },
      {
        name: 'thirdColumn',
        sample: 0.25,
        source: 'source3',
        validity: '2023',
        isGeoCode: false,
      },
      {
        name: 'fourthColumn',
        sample: 1000,
        source: 'source4',
        validity: '2024',
        isGeoCode: false,
      },
    ],
    dataset: mockDataset,
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
];

export const indicatorsMockData: ApiData<Indicator>[] = [
  {
    id: 4,
    name: 'mockIndicator1',
    description: 'This is a mock indicator',
    source: 'mock',
    validity: 2,
    created_at: '2023-03-14T10:06:41.979Z',
    updated_at: '2023-03-14T10:06:41.979Z',
    equation: 'A+2',
    variables: [],
  },
  {
    id: 5,
    name: 'mockIndicator2',
    description: 'This is a mock indicator',
    source: 'mock',
    validity: 4,
    variables: [],
    created_at: '2023-03-14T10:12:37.285Z',
    updated_at: '2023-03-14T10:12:37.285Z',
    equation: 'A*2',
  },
];

export const mockPageCartoData: ApiData<PageCarto> = {
  id: 1,
  name: 'Page Carto 1',
  owner: mockUser,
  tags: mockTags,
  map: mockGeoMap,
  cover: mockCover,
  indicators: indicatorsMockData,
  charts: [],
  html: '<p>Test HTML</p>',
  data_fragments: mockDataFragments,
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockGeoMapData: ApiData<GeoMap> = {
  id: 1,
  name: 'Test Map',
  yearValidity: 2022,
  mesh: 'Test Maille',
  source: 'Test Source',
  license: 'Test License',
  properties: [
    {
      name: 'nom',
      sample: 'Taponas',
      isGeoCode: true,
    },
  ],
  owner: mockUser,
  geoJSON: {
    id: 1,
    name: 'test.geojson',
    hash: 'abc123',
    provider: 'provider123',
    ext: '.geojson',
    mime: 'application/json',
    size: 1234,
    url: 'https://example.com/api/v1/geo-maps/id/thumbnail',
    formats: {
      '720': {
        extension: 'ext',
        url: 'happyUrl',
        size: 360,
        hash: 'hashbar',
        mime: 'mimeExample',
      },
    },
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockGeoMapApiCollection: ApiCollection<GeoMap> = {
  data: Array.from({ length: 3 }, (_, index) => ({
    ...mockGeoMapData,
    id: index + 1,
    name: `Test Map ${index + 1}`,
  })),
  meta: {
    pagination: {
      pageCount: 1,
      total: 3,
    },
  },
};
