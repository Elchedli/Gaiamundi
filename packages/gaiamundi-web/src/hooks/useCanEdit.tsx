import { useLocation } from 'react-router-dom';
import { useAuth } from './useAuth';

export const useCanEdit = () => {
  const { isAuthenticated } = useAuth();
  const location = useLocation();
  const isOnEditorPage = location.pathname.endsWith('edit');
  return isAuthenticated && isOnEditorPage;
};
