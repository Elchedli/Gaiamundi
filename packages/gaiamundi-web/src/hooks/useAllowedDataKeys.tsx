import { useMemo } from 'react';
import { ChartType, RawDatumType } from '../interfaces/chart';
import { useChartConfig } from './useChartConfig';
import { useData } from './useData';

export const ALLOWED_DOMAIN_TYPES_BY_CHART: {
  [key in ChartType]: Array<RawDatumType>;
} = {
  column: ['string', 'number'],
  bar: ['string', 'number'],
  scatter: ['number'],
  line: ['number'],
  lineColumn: ['string', 'number'],
  pie: ['number'],
  area: ['number'],
  bubble: ['number'],
};

export const useAllowedDataKeys = () => {
  const { chart: chartConfig } = useChartConfig();
  const { dataKeys } = useData();
  const allowedTypes = ALLOWED_DOMAIN_TYPES_BY_CHART[chartConfig.type];

  const domainKeyOptions = useMemo(
    () =>
      Object.entries(dataKeys)
        .filter(([, type]) => {
          return allowedTypes.indexOf(type) !== -1;
        })
        .map(([key]) => ({
          value: key,
          label: key,
        })),
    [allowedTypes, dataKeys]
  );

  return { domainKeyOptions };
};
