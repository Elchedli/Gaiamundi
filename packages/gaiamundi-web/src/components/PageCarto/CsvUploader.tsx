import { useState } from 'react';
import { useMutation } from 'react-query';

import { useToast } from 'hooks/useToast';
import { UploadedFile } from 'interfaces/file';
import LoadingSpinner from 'components/Icons/LoadingSpinner';
import { ApiErrorAlert } from 'components/Alert/ApiErrorMessage';
import { ApiError } from 'interfaces/api';
import DropZone from '../Forms/Inputs/DropZone';
import { validateCsv } from 'utils/file';
import {
  DocumentChartBarIcon,
  ExclamationCircleIcon,
} from '@heroicons/react/24/solid';
import { uploadCsv } from 'services/page-carto';
import { Alert } from 'components/Alert/Alert';
import { Button } from 'components/Button/Button';

type CsvUploaderProps = {
  onFileUploaded: (file: UploadedFile) => void;
  onFileParse?: (columnNames: string[] | undefined) => void;
  onCancel: () => void;
};

export const CsvUploader: React.FC<CsvUploaderProps> = ({
  onFileUploaded,
  onFileParse,
  onCancel,
}) => {
  const { addToast } = useToast();
  const [file, setFile] = useState<UploadedFile | undefined>(undefined);
  const [errors, setErrors] = useState<string>('');

  const { mutateAsync, isError, isLoading, error } = useMutation({
    mutationFn: async (data: { file: File }) => {
      return await uploadCsv(data.file);
    },
    onSuccess: (data: UploadedFile) => {
      setFile(data);
      onFileUploaded(data);
      addToast({
        title: `Fichier téléchargé`,
        description: `Fichier ${data.name} téléchargé avec succès`,
        type: 'success',
      });
    },
  });

  const onUpload = async (file: File) => {
    setErrors('');
    try {
      const data = await validateCsv(file);
      onFileParse && onFileParse(data.meta.fields);
      if (Array.isArray(data.errors) && data.errors.length > 0) {
        console.warn(`CSV Parse errors : `, data.errors);
      }
      // This is done to set the displayed filename after the file is uploaded
      if (data.errors.length > 0) {
        setErrors(
          `Le fichier CSV contient des erreurs dans les lignes : ${data.errors
            .map((e) => e.row)
            .join(', ')}`
        );
      }
      mutateAsync({ file });
    } catch (e) {
      addToast({
        title: `Erreur lors de la validation du fichier CSV`,
        description: (e as Error).message,
        type: 'error',
      });
    }
  };

  const handleCancel = () => {
    setFile(undefined);
    setErrors('');
    onCancel();
  };

  if (isLoading) {
    return <LoadingSpinner />;
  }

  if (isError) {
    <ApiErrorAlert error={error as ApiError} />;
  }

  return (
    <div className="self-center">
      <h2>{`Uploader des données à partir d'un fichier CSV`}</h2>
      {file ? (
        <div className="flex flex-col items-center justify-center pt-5 pb-6">
          <DocumentChartBarIcon width={64} height={64} />
          <span className="mb-2 text-sm text-gray-500 dark:text-gray-400">
            {file.name} ({file.mime})
          </span>
          <Button color="red" onClick={handleCancel} size="xs">
            Annuler
          </Button>
        </div>
      ) : (
        <DropZone onUpload={onUpload} />
      )}
      {errors && (
        <Alert className="p-2 text-sm" icon={ExclamationCircleIcon}>
          {errors}
        </Alert>
      )}
    </div>
  );
};
