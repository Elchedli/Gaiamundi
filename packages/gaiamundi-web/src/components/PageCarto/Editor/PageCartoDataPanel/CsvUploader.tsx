import { useState } from 'react';
import { useMutation } from 'react-query';

import {
  DocumentChartBarIcon,
  ExclamationCircleIcon,
} from '@heroicons/react/24/solid';

import { Alert } from 'components/Alert/Alert';
import { ApiErrorAlert } from 'components/Alert/ApiErrorMessage';
import { Button } from 'components/Button/Button';
import LoadingSpinner from 'components/Icons/LoadingSpinner';
import DropZone from 'components/Inputs/DropZone';
import { useToast } from 'hooks/useToast';
import { ApiError } from 'interfaces/api';
import { Column } from 'interfaces/column';
import { UploadedFile } from 'interfaces/file';
import { uploadCsv } from 'services/page-carto';
import { parseCsvColumns, validateCsv } from 'utils/file';

type CsvUploaderProps = {
  onUpload: (file: UploadedFile) => void;
  onParse?: (columnNames: Column[]) => void;
  onCancel?: () => void;
  onChange: (fileId: number) => void;
};

export const CsvUploader: React.FC<CsvUploaderProps> = ({
  onUpload,
  onParse,
  onCancel,
  onChange,
}) => {
  const { addToast } = useToast();
  const [file, setFile] = useState<UploadedFile | undefined>(undefined);
  const [errors, setErrors] = useState<string>('');

  const { mutateAsync, isError, isLoading, error } = useMutation({
    mutationFn: async (data: { file: File }) => {
      return await uploadCsv(data.file);
    },
    onSuccess: (data: UploadedFile) => {
      setFile(data);
      onUpload(data);
      onChange(data.id);
      addToast({
        title: `Fichier téléchargé`,
        description: `Fichier ${data.name} téléchargé avec succès`,
        type: 'success',
      });
    },
  });

  const handleUpload = async (csvFile: File) => {
    setErrors('');
    try {
      const data = await validateCsv(csvFile);
      if (onParse) {
        const columns = parseCsvColumns(data.meta.fields || []);
        onParse(columns);
      }

      if (Array.isArray(data.errors) && data.errors.length > 0) {
        // eslint-disable-next-line no-console
        console.warn(`CSV Parse errors : `, data.errors);
      }
      // This is done to set the displayed filename after the file is uploaded
      if (data.errors.length > 0) {
        setErrors(
          `Le fichier CSV contient des erreurs dans les lignes : ${data.errors
            .map((e) => e.row)
            .join(', ')}`
        );
      }
      mutateAsync({ file: csvFile });
    } catch (e) {
      addToast({
        title: `Erreur lors de la validation du fichier CSV`,
        description: (e as Error).message,
        type: 'error',
      });
    }
  };

  const handleCancel = () => {
    setFile(undefined);
    setErrors('');
    onCancel && onCancel();
  };

  if (isLoading) {
    return <LoadingSpinner />;
  }

  if (isError) {
    <ApiErrorAlert error={error as ApiError} />;
  }

  return (
    <div className="self-center">
      {file ? (
        <div className="flex flex-col items-center justify-center pt-5 pb-6">
          <DocumentChartBarIcon width={64} height={64} />
          <span className="mb-2 text-sm text-gray-500 dark:text-gray-400">
            {file.name} ({file.mime})
          </span>
          <Button color="red" onClick={handleCancel} size="xs">
            Annuler
          </Button>
        </div>
      ) : (
        <DropZone onUpload={handleUpload} />
      )}
      {errors && (
        <Alert className="p-2 text-sm" icon={ExclamationCircleIcon}>
          {errors}
        </Alert>
      )}
    </div>
  );
};
