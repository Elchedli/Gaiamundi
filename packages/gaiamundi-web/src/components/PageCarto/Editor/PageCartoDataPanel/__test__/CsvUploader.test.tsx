import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useAuth } from 'hooks/useAuth';
import { QueryClient, QueryClientProvider } from 'react-query';
import { mockUser } from 'utils/mocks/data';
import { CsvUploader } from '../CsvUploader';

jest.mock('hooks/useAuth');

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

describe('CsvUploader', () => {
  beforeEach(() => {
    (useAuth as jest.Mock).mockImplementation(() => ({
      isAuthenticated: true,
      user: mockUser,
    }));
  });

  const onUploadMock = jest.fn();
  const onParseMock = jest.fn();
  const onCancelMock = jest.fn();
  const onChangeMock = jest.fn();

  it('should upload a CSV file and click  on cancel button when upload is successful', async () => {
    const file = new File(['a,b,c\n1,2,3'], 'file.csv', { type: 'text/csv' });
    const { container } = render(
      <QueryClientProvider client={queryClient}>
        <CsvUploader
          onUpload={onUploadMock}
          onParse={onParseMock}
          onCancel={onCancelMock}
          onChange={onChangeMock}
        />
      </QueryClientProvider>
    );
    const fileInput = await screen.findByTestId('hidden-file-input');
    userEvent.upload(fileInput, file);
    const loadedFilePage = await screen.findByTestId('loadedFilePage');
    expect(loadedFilePage).toBeInTheDocument();
    userEvent.click(loadedFilePage.querySelector('button') as HTMLElement);
    expect(container.contains(loadedFilePage)).not.toBeTruthy();
  });
});
