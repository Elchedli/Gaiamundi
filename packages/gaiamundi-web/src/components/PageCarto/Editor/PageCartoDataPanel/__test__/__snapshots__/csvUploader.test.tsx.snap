// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CsvUploader should upload a CSV file 1`] = `
<div>
  <div
    class="self-center"
  >
    <div
      data-testid="dropzone"
    >
      <label
        class="text-sm font-medium text-gray-900 dark:text-gray-300 flex flex-col border-2 border-gray-300 border-dashed rounded-lg cursor-pointer bg-gray-50 dark:hover:bg-bray-800 dark:bg-gray-700 hover:bg-gray-100 dark:border-gray-600 dark:hover:border-gray-500 dark:hover:bg-gray-600"
        for="dropzone-file"
      >
        <div
          class="flex flex-col items-center justify-center pt-5 pb-6"
        >
          <svg
            aria-hidden="true"
            class="w-10 h-10 mb-3 text-gray-400"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 16a4 4 0 01-.88-7.903A5 5 0 1115.9 6L16 6a5 5 0 011 9.9M15 13l-3-3m0 0l-3 3m3-3v12"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
          <p
            class="mb-2 text-sm text-gray-500 dark:text-gray-400"
          >
            <span>
              <span
                class="font-semibold"
              >
                Cliquer pour ajouter 
              </span>
              ou glisser et déposer
            </span>
          </p>
        </div>
        <div>
          <div
            class="relative"
          >
            <input
              aria-describedby="file_upload_help"
              class="border disabled:cursor-not-allowed disabled:opacity-50 bg-gray-50 border-gray-300 text-gray-900 focus:border-blue-500 focus:ring-blue-500 dark:border-gray-600 dark:bg-gray-700 dark:text-white dark:placeholder-gray-400 dark:focus:border-blue-500 dark:focus:ring-blue-500 rounded-lg p-2.5 text-sm hidden"
              data-testid="hidden-file-input"
              id="dropzone-file"
              type="file"
            />
          </div>
        </div>
      </label>
    </div>
  </div>
</div>
`;

exports[`csvUploader renders error message 1`] = `
<div>
  <div
    class="self-center"
  >
    <div
      data-testid="dropzone"
    >
      <label
        class="text-sm font-medium text-gray-900 dark:text-gray-300 flex flex-col border-2 border-gray-300 border-dashed rounded-lg cursor-pointer bg-gray-50 dark:hover:bg-bray-800 dark:bg-gray-700 hover:bg-gray-100 dark:border-gray-600 dark:hover:border-gray-500 dark:hover:bg-gray-600"
        for="dropzone-file"
      >
        <div
          class="flex flex-col items-center justify-center pt-5 pb-6"
        >
          <svg
            aria-hidden="true"
            class="w-10 h-10 mb-3 text-gray-400"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 16a4 4 0 01-.88-7.903A5 5 0 1115.9 6L16 6a5 5 0 011 9.9M15 13l-3-3m0 0l-3 3m3-3v12"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
          <p
            class="mb-2 text-sm text-gray-500 dark:text-gray-400"
          >
            <span>
              <span
                class="font-semibold"
              >
                Cliquer pour ajouter 
              </span>
              ou glisser et déposer
            </span>
          </p>
        </div>
        <div>
          <div
            class="relative"
          >
            <input
              aria-describedby="file_upload_help"
              class="border disabled:cursor-not-allowed disabled:opacity-50 bg-gray-50 border-gray-300 text-gray-900 focus:border-blue-500 focus:ring-blue-500 dark:border-gray-600 dark:bg-gray-700 dark:text-white dark:placeholder-gray-400 dark:focus:border-blue-500 dark:focus:ring-blue-500 rounded-lg p-2.5 text-sm hidden"
              data-testid="hidden-file-input"
              id="dropzone-file"
              type="file"
            />
          </div>
        </div>
      </label>
    </div>
  </div>
</div>
`;
